File: threads/src/main/java/im/threads/business/config/BaseConfig.java
package im.threads.business.config;

import android.content.Context;
import android.net.Uri;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.security.KeyStore;
import java.util.List;

import javax.net.ssl.SSLSocketFactory;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;

import im.threads.business.core.UnreadMessagesCountListener;
import im.threads.business.exceptions.MetaConfigurationException;
import im.threads.business.imageLoading.ImageLoaderOkHttpProvider;
import im.threads.business.logger.core.LoggerConfig;
import im.threads.business.models.SslSocketFactoryConfig;
import im.threads.business.rest.config.RequestConfig;
import im.threads.business.rest.config.SocketClientSettings;
import im.threads.business.transport.Transport;
import im.threads.business.transport.threadsGate.ThreadsGateTransport;
import im.threads.business.utils.MetadataBusiness;
import im.threads.business.utils.TlsConfigurationUtils;
import im.threads.business.utils.gson.UriDeserializer;
import im.threads.business.utils.gson.UriSerializer;
import okhttp3.Interceptor;

public class BaseConfig {

    private static final String TAG = BaseConfig.class.getSimpleName();
    public static BaseConfig instance;

    @NonNull
    public final Context context;

    public final RequestConfig requestConfig;
    public final SslSocketFactoryConfig sslSocketFactoryConfig;

    @Nullable
    public final Interceptor networkInterceptor;
    @Nullable
    public final UnreadMessagesCountListener unreadMessagesCountListener;
    @NonNull
    public final Transport transport;
    @NonNull
    public final String serverBaseUrl;
    @NonNull
    public final String datastoreUrl;

    public final boolean isDebugLoggingEnabled;
    /**
     * set history loading count
     */
    public final int historyLoadingCount;

    public final int surveyCompletionDelay;

    public final boolean newChatCenterApi;

    @Nullable
    public final LoggerConfig loggerConfig;

    public final Gson gson = new GsonBuilder()
            .registerTypeAdapter(Uri.class, new UriSerializer())
            .registerTypeAdapter(Uri.class, new UriDeserializer())
            .create();

    public BaseConfig(@NonNull Context context,
                      @Nullable String serverBaseUrl,
                      @Nullable String datastoreUrl,
                      @Nullable String threadsGateUrl,
                      @Nullable String threadsGateProviderUid,
                      @Nullable String threadsGateHCMProviderUid,
                      @Nullable Boolean isNewChatCenterApi,
                      @Nullable LoggerConfig loggerConfig,
                      @Nullable UnreadMessagesCountListener unreadMessagesCountListener,
                      @Nullable Interceptor networkInterceptor,
                      boolean isDebugLoggingEnabled,
                      int historyLoadingCount,
                      int surveyCompletionDelay,
                      @NonNull RequestConfig requestConfig,
                      List<Integer> certificateRawResIds) {
        this.context = context.getApplicationContext();
        this.unreadMessagesCountListener = unreadMessagesCountListener;
        this.networkInterceptor = networkInterceptor;
        this.isDebugLoggingEnabled = isDebugLoggingEnabled;
        this.newChatCenterApi = getIsNewChatCenterApi(isNewChatCenterApi);
        this.loggerConfig = loggerConfig;
        this.historyLoadingCount = historyLoadingCount;
        this.surveyCompletionDelay = surveyCompletionDelay;
        this.sslSocketFactoryConfig = getSslSocketFactoryConfig(certificateRawResIds);
        this.transport = getTransport(threadsGateUrl, threadsGateProviderUid,
                threadsGateHCMProviderUid, requestConfig.getSocketClientSettings());
        this.serverBaseUrl = getServerBaseUrl(serverBaseUrl);
        this.datastoreUrl = getDatastoreUrl(datastoreUrl);
        this.requestConfig = requestConfig;
        ImageLoaderOkHttpProvider.INSTANCE.createOkHttpClient(
                requestConfig.getPicassoHttpClientSettings(),
                sslSocketFactoryConfig
        );
    }

    private SslSocketFactoryConfig getSslSocketFactoryConfig(List<Integer> certificateRawResIds) {
        if (certificateRawResIds == null || certificateRawResIds.isEmpty())
            return null;

        KeyStore keyStore = TlsConfigurationUtils.createTlsPinningKeyStore(
                context.getResources(),
                certificateRawResIds
        );
        TrustManager[] trustManagers = TlsConfigurationUtils.getTrustManagers(keyStore);
        X509TrustManager trustManager = TlsConfigurationUtils.getX509TrustManager(trustManagers);
        SSLSocketFactory sslSocketFactory =
                TlsConfigurationUtils.createTlsPinningSocketFactory(trustManagers);
        return new SslSocketFactoryConfig(sslSocketFactory, trustManager);
    }

    private Transport getTransport(@Nullable String providedThreadsGateUrl,
                                   @Nullable String providedThreadsGateProviderUid,
                                   @Nullable String providedThreadsGateHCMProviderUid,
                                   SocketClientSettings socketClientSettings) {
        String threadsGateProviderUid = !TextUtils.isEmpty(providedThreadsGateProviderUid)
                ? providedThreadsGateProviderUid
                : MetadataBusiness.getThreadsGateProviderUid(this.context);
        String threadsGateHCMProviderUid = !TextUtils.isEmpty(providedThreadsGateHCMProviderUid)
                ? providedThreadsGateHCMProviderUid
                : MetadataBusiness.getThreadsGateHCMProviderUid(this.context);
        String threadsGateUrl = !TextUtils.isEmpty(providedThreadsGateUrl)
                ? providedThreadsGateUrl
                : MetadataBusiness.getThreadsGateUrl(this.context);
        if (TextUtils.isEmpty(threadsGateUrl)) {
            throw new MetaConfigurationException("Threads gate url is not set");
        }
        if (TextUtils.isEmpty(threadsGateProviderUid)) {
            throw new MetaConfigurationException("Threads gate provider uid is not set");
        }
        return new ThreadsGateTransport(
                threadsGateUrl,
                threadsGateProviderUid,
                threadsGateHCMProviderUid,
                isDebugLoggingEnabled,
                socketClientSettings,
                sslSocketFactoryConfig,
                networkInterceptor
        );
    }

    @NonNull
    private String getServerBaseUrl(@Nullable String serverBaseUrl) {
        String baseUrl = TextUtils.isEmpty(serverBaseUrl) ? MetadataBusiness.getServerBaseUrl(this.context) : serverBaseUrl;
        if (baseUrl == null) {
            throw new MetaConfigurationException("Neither im.threads.getServerUrl meta variable, nor serverBaseUrl were provided");
        }
        if (!baseUrl.endsWith("/")) {
            baseUrl = baseUrl + "/";
        }
        return baseUrl;
    }

    @NonNull
    private String getDatastoreUrl(@Nullable String dataStoreUrl) {
        String datastoreUrl = TextUtils.isEmpty(dataStoreUrl) ? MetadataBusiness.getDatastoreUrl(this.context) : dataStoreUrl;
        if (datastoreUrl == null) {
            throw new MetaConfigurationException("Neither im.threads.getDatastoreUrl meta variable, nor datastoreUrl were provided");
        }
        if (!datastoreUrl.endsWith("/")) {
            datastoreUrl = datastoreUrl + "/";
        }
        return datastoreUrl;
    }

    private boolean getIsNewChatCenterApi(@Nullable Boolean isNewChatCenterApi) {
        return isNewChatCenterApi == null ? MetadataBusiness.getNewChatCenterApi(this.context) : isNewChatCenterApi;
    }
}

File: threads/src/main/java/im/threads/business/config/BaseConfig.kt
package im.threads.business.config

import android.annotation.SuppressLint
import android.content.Context
import android.net.Uri
import com.google.gson.GsonBuilder
import im.threads.business.core.UnreadMessagesCountListener
import im.threads.business.exceptions.MetaConfigurationException
import im.threads.business.imageLoading.ImageLoaderOkHttpProvider
import im.threads.business.logger.LoggerConfig
import im.threads.business.models.SslSocketFactoryConfig
import im.threads.business.models.enums.ApiVersionEnum
import im.threads.business.rest.config.RequestConfig
import im.threads.business.rest.config.SocketClientSettings
import im.threads.business.serviceLocator.core.inject
import im.threads.business.transport.Transport
import im.threads.business.transport.threadsGate.ThreadsGateTransport
import im.threads.business.utils.createSslSocketFactoryConfig
import im.threads.business.utils.createTlsPinningKeyStore
import im.threads.business.utils.getTrustManagers
import im.threads.business.utils.gson.UriDeserializer
import im.threads.business.utils.gson.UriSerializer
import okhttp3.Interceptor
import java.security.cert.X509Certificate
import javax.net.ssl.TrustManager
import javax.net.ssl.X509TrustManager

open class BaseConfig(
    context: Context,
    var serverBaseUrl: String?,
    var datastoreUrl: String?,
    var threadsGateUrl: String?,
    var threadsGateProviderUid: String?,
    var isNewChatCenterApi: Boolean = false,
    val loggerConfig: LoggerConfig?,
    val unreadMessagesCountListener: UnreadMessagesCountListener?,
    val networkInterceptor: Interceptor?,
    val isDebugLoggingEnabled: Boolean,
    val historyLoadingCount: Int,
    val surveyCompletionDelay: Int,
    val requestConfig: RequestConfig,
    val notificationImportance: Int,
    var trustedSSLCertificates: List<Int>?,
    var allowUntrustedSSLCertificate: Boolean,
    var keepSocketActive: Boolean,
    var apiVersion: ApiVersionEnum
) {
    @JvmField val context: Context
    var sslSocketFactoryConfig: SslSocketFactoryConfig?

    @JvmField
    var transport: Transport

    @JvmField
    val gson = GsonBuilder()
        .registerTypeAdapter(Uri::class.java, UriSerializer())
        .registerTypeAdapter(Uri::class.java, UriDeserializer())
        .create()

    private val imageLoaderOkHttpProvider: ImageLoaderOkHttpProvider by inject()

    init {
        this.context = context.applicationContext
        transport = getTransport(threadsGateUrl, threadsGateProviderUid, requestConfig.socketClientSettings)
        sslSocketFactoryConfig = getSslSocketFactoryConfig(trustedSSLCertificates)
        imageLoaderOkHttpProvider.createOkHttpClient(
            requestConfig.picassoHttpClientSettings,
            sslSocketFactoryConfig
        )
    }

    internal fun updateTransport(
        threadsGateUrl: String,
        threadsGateProviderUid: String,
        trustedSSLCertificates: List<Int>?
    ) {
        this.trustedSSLCertificates = trustedSSLCertificates
        sslSocketFactoryConfig = getSslSocketFactoryConfig(trustedSSLCertificates)
        imageLoaderOkHttpProvider.createOkHttpClient(
            requestConfig.picassoHttpClientSettings,
            sslSocketFactoryConfig
        )
        transport = getTransport(threadsGateUrl, threadsGateProviderUid, requestConfig.socketClientSettings)
    }

    private fun getSslSocketFactoryConfig(trustedSSLCertificates: List<Int>?): SslSocketFactoryConfig? {
        if (!trustedSSLCertificates.isNullOrEmpty()) {
            val keyStore = createTlsPinningKeyStore(
                context.resources,
                trustedSSLCertificates
            )
            return createSslSocketFactoryConfig(getTrustManagers(keyStore))
        }
        if (allowUntrustedSSLCertificate) {
            @SuppressLint("CustomX509TrustManager", "TrustAllX509TrustManager")
            val trustManagers = arrayOf<TrustManager>(object : X509TrustManager {
                @SuppressLint("TrustAllX509TrustManager")
                override fun checkClientTrusted(chain: Array<X509Certificate>, authType: String) {}

                @SuppressLint("TrustAllX509TrustManager")
                override fun checkServerTrusted(chain: Array<X509Certificate>, authType: String) {}

                override fun getAcceptedIssuers(): Array<X509Certificate> {
                    return arrayOf()
                }
            })
            return createSslSocketFactoryConfig(trustManagers)
        }
        return null
    }

    private fun getTransport(
        providedThreadsGateUrl: String?,
        providedThreadsGateProviderUid: String?,
        socketClientSettings: SocketClientSettings
    ): Transport {
        if (providedThreadsGateUrl.isNullOrBlank()) {
            throw MetaConfigurationException("Threads gate url is not set")
        }
        if (providedThreadsGateProviderUid.isNullOrBlank()) {
            throw MetaConfigurationException("Threads gate provider uid is not set")
        }
        return ThreadsGateTransport(
            providedThreadsGateUrl,
            providedThreadsGateProviderUid,
            isDebugLoggingEnabled,
            socketClientSettings,
            sslSocketFactoryConfig,
            networkInterceptor,
            context
        )
    }

    companion object {
        @SuppressLint("StaticFieldLeak")
        private var baseInstance: BaseConfig? = null

        fun getInstance() = baseInstance!!

        fun setInstance(instance: BaseConfig) {
            baseInstance = instance
        }
    }
}
